File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/MapProjectionHandle.java
package org.maplibre.nativeffi;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.util.List;
import java.util.Objects;
import org.maplibre.nativeffi.internal.HandleState;
import org.maplibre.nativeffi.internal.MemoryUtil;
import org.maplibre.nativeffi.internal.NativeAccess;
import org.maplibre.nativeffi.internal.Status;
import org.maplibre.nativeffi.internal.Structs;
import org.maplibre.nativeffi.internal.c.MapLibreNativeC;
import org.maplibre.nativeffi.internal.c.mln_lat_lng;
import org.maplibre.nativeffi.internal.c.mln_screen_point;

/** Owned standalone projection snapshot created from a map. */
public final class MapProjectionHandle implements AutoCloseable {
  private final HandleState state;

  private MapProjectionHandle(MemorySegment handle) {
    this.state = new HandleState("MapProjectionHandle", handle);
  }

  public static MapProjectionHandle create(MapHandle map) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(map, "map");
    try (var arena = Arena.ofConfined()) {
      var outProjection = MemoryUtil.allocatePointer(arena);
      Status.check(MapLibreNativeC.mln_map_projection_create(map.nativeHandle(), outProjection));
      return new MapProjectionHandle(outProjection.get(ValueLayout.ADDRESS, 0));
    }
  }

  public CameraOptions camera() {
    NativeAccess.ensureLoaded();
    try (var arena = Arena.ofConfined()) {
      var outCamera = MapLibreNativeC.mln_camera_options_default(arena);
      Status.check(MapLibreNativeC.mln_map_projection_get_camera(state.requireLive(), outCamera));
      return Structs.cameraOptions(outCamera);
    }
  }

  public void setCamera(CameraOptions camera) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(camera, "camera");
    try (var arena = Arena.ofConfined()) {
      Status.check(
          MapLibreNativeC.mln_map_projection_set_camera(
              state.requireLive(), Structs.cameraOptions(camera, arena)));
    }
  }

  public void setVisibleCoordinates(List<LatLng> coordinates, EdgeInsets padding) {
    NativeAccess.ensureLoaded();
    var copiedCoordinates = List.copyOf(Objects.requireNonNull(coordinates, "coordinates"));
    if (copiedCoordinates.isEmpty()) {
      throw new IllegalArgumentException("coordinates must not be empty");
    }
    Objects.requireNonNull(padding, "padding");
    try (var arena = Arena.ofConfined()) {
      Status.check(
          MapLibreNativeC.mln_map_projection_set_visible_coordinates(
              state.requireLive(),
              Structs.latLngArray(copiedCoordinates, arena),
              copiedCoordinates.size(),
              Structs.edgeInsets(padding, arena)));
    }
  }

  public ScreenPoint pixelForLatLng(LatLng coordinate) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(coordinate, "coordinate");
    try (var arena = Arena.ofConfined()) {
      var outPoint = mln_screen_point.allocate(arena);
      Status.check(
          MapLibreNativeC.mln_map_projection_pixel_for_lat_lng(
              state.requireLive(), Structs.latLng(coordinate, arena), outPoint));
      return Structs.screenPoint(outPoint);
    }
  }

  public LatLng latLngForPixel(ScreenPoint point) {
    NativeAccess.ensureLoaded();
    Objects.requireNonNull(point, "point");
    try (var arena = Arena.ofConfined()) {
      var outCoordinate = mln_lat_lng.allocate(arena);
      Status.check(
          MapLibreNativeC.mln_map_projection_lat_lng_for_pixel(
              state.requireLive(), Structs.screenPoint(point, arena), outCoordinate));
      return Structs.latLng(outCoordinate);
    }
  }

  @Override
  public void close() {
    NativeAccess.ensureLoaded();
    state.closeOnce(MapLibreNativeC::mln_map_projection_destroy);
  }

  public boolean isClosed() {
    return state.isReleased();
  }
}

File: bindings/kotlin-native/src/nativeMain/kotlin/org/maplibre/nativeffi/map/MapProjectionHandle.kt
package org.maplibre.nativeffi.map

import cnames.structs.mln_map_projection
import kotlinx.cinterop.CPointer
import kotlinx.cinterop.CPointerVarOf
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.alloc
import kotlinx.cinterop.memScoped
import kotlinx.cinterop.pointed
import kotlinx.cinterop.ptr
import kotlinx.cinterop.value
import org.maplibre.nativeffi.camera.CameraOptions
import org.maplibre.nativeffi.internal.c.mln_camera_options_default
import org.maplibre.nativeffi.internal.c.mln_map_projection_create
import org.maplibre.nativeffi.internal.c.mln_map_projection_destroy
import org.maplibre.nativeffi.internal.c.mln_map_projection_get_camera
import org.maplibre.nativeffi.internal.c.mln_map_projection_set_camera
import org.maplibre.nativeffi.internal.lifecycle.HandleState
import org.maplibre.nativeffi.internal.status.Status
import org.maplibre.nativeffi.internal.struct.MapStructs

/** Owned standalone projection snapshot created from a map. */
@OptIn(ExperimentalForeignApi::class)
public class MapProjectionHandle private constructor(handle: CPointer<mln_map_projection>) :
  AutoCloseable {
  private val state = HandleState("MapProjectionHandle", handle)

  public fun camera(): CameraOptions = memScoped {
    val outCamera = mln_camera_options_default().getPointer(this)
    Status.check(mln_map_projection_get_camera(state.requireLive(), outCamera))
    MapStructs.cameraOptions(outCamera.pointed)
  }

  public fun setCamera(camera: CameraOptions) {
    memScoped {
      Status.check(
        mln_map_projection_set_camera(state.requireLive(), MapStructs.cameraOptions(camera, this))
      )
    }
  }

  override fun close() {
    state.closeOnce(::mln_map_projection_destroy)
  }

  public fun isClosed(): Boolean = state.isReleased()

  internal fun nativeHandle(): CPointer<mln_map_projection> = state.requireLive()

  internal fun nativeAddress(): Long = state.address()

  public companion object {
    public fun create(map: MapHandle): MapProjectionHandle = memScoped {
      val outProjection = alloc<CPointerVarOf<CPointer<mln_map_projection>>>()
      outProjection.value = null
      Status.check(mln_map_projection_create(map.nativeHandle(), outProjection.ptr))
      MapProjectionHandle(
        requireNotNull(outProjection.value) { "mln_map_projection_create returned null" }
      )
    }
  }
}
